SQL: Exception with OR of impossible filters

### Description

A query like this, which has and OR of two impossible time filters (a floor to day can never be any hour other than "00:00:00"):

```sql
SELECT COUNT(*) FROM wikiticker WHERE FLOOR(__time TO DAY) = TIMESTAMP '2000-01-02 01:00:00' OR FLOOR(__time TO DAY) = TIMESTAMP '2000-01-02 02:00:00'
```

Gives an error like the following:

```
java.lang.RuntimeException: Error while applying rule DruidQueryRule(WHERE_FILTER), args [rel#30:LogicalFilter.NONE.[](input=rel#19:Subset#0.NONE.[],condition=OR(=(TIME_FLOOR($0, 'P1M', null, 'America/Los_Angeles'), 2000-01-01 00:00:00), =(TIME_FLOOR($0, 'P1M', null, 'America/Los_Angeles'), 2000-02-01 00:00:00))), rel#43:DruidQueryRel.NONE.[](query={"queryType":"scan","dataSource":{"type":"table","name":"foo"},"intervals":{"type":"intervals","intervals":["-146136543-09-08T08:23:32.096Z/146140482-04-24T15:36:27.903Z"]},"virtualColumns":[],"resultFormat":"compactedList","batchSize":20480,"limit":9223372036854775807,"order":"none","filter":null,"columns":["__time","cnt","dim1","dim2","dim3","m1","m2","unique_dim1"],"legacy":false,"context":{"defaultTimeout":300000,"maxScatterGatherBytes":9223372036854775807,"sqlCurrentTimestamp":"2000-01-01T00:00:00Z","sqlQueryId":"dummy"},"descending":false,"granularity":{"type":"all"}},signature={__time:LONG, cnt:LONG, dim1:STRING, dim2:STRING, dim3:STRING, m1:FLOAT, m2:DOUBLE, unique_dim1:COMPLEX})]

	at org.apache.calcite.plan.volcano.VolcanoRuleCall.onMatch(VolcanoRuleCall.java:236)
	at org.apache.calcite.plan.volcano.VolcanoPlanner.findBestExp(VolcanoPlanner.java:646)
	at org.apache.calcite.tools.Programs$RuleSetProgram.run(Programs.java:339)
	at org.apache.calcite.tools.Programs$SequenceProgram.run(Programs.java:358)
	at org.apache.calcite.prepare.PlannerImpl.transform(PlannerImpl.java:337)
	at org.apache.druid.sql.calcite.planner.DruidPlanner.planWithDruidConvention(DruidPlanner.java:119)
	at org.apache.druid.sql.calcite.planner.DruidPlanner.plan(DruidPlanner.java:89)
	at org.apache.druid.sql.SqlLifecycle.plan(SqlLifecycle.java:143)
	at org.apache.druid.sql.SqlLifecycle.planAndAuthorize(SqlLifecycle.java:209)
	at org.apache.druid.sql.SqlLifecycle.runSimple(SqlLifecycle.java:246)
	at org.apache.druid.sql.calcite.BaseCalciteQueryTest.getResults(BaseCalciteQueryTest.java:571)
	at org.apache.druid.sql.calcite.BaseCalciteQueryTest.getResults(BaseCalciteQueryTest.java:518)
	at org.apache.druid.sql.calcite.BaseCalciteQueryTest.testQuery(BaseCalciteQueryTest.java:507)
	at org.apache.druid.sql.calcite.BaseCalciteQueryTest.testQuery(BaseCalciteQueryTest.java:458)
	at org.apache.druid.sql.calcite.CalciteQueryTest.testGroupAndFilterOnTimeFloorWithTimeZone(CalciteQueryTest.java:5289)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.apache.druid.sql.calcite.util.QueryLogHook$1.evaluate(QueryLogHook.java:95)
	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:239)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: java.lang.IllegalStateException: newChildren.size > 0
	at com.google.common.base.Preconditions.checkState(Preconditions.java:176)
	at org.apache.druid.sql.calcite.filtration.CombineAndSimplifyBounds.doSimplify(CombineAndSimplifyBounds.java:203)
	at org.apache.druid.sql.calcite.filtration.CombineAndSimplifyBounds.doSimplifyOr(CombineAndSimplifyBounds.java:114)
	at org.apache.druid.sql.calcite.filtration.CombineAndSimplifyBounds.process(CombineAndSimplifyBounds.java:62)
	at org.apache.druid.sql.calcite.filtration.BottomUpTransform.checkedProcess(BottomUpTransform.java:38)
	at org.apache.druid.sql.calcite.filtration.BottomUpTransform.apply0(BottomUpTransform.java:81)
	at org.apache.druid.sql.calcite.filtration.BottomUpTransform.apply(BottomUpTransform.java:46)
	at org.apache.druid.sql.calcite.filtration.BottomUpTransform.apply(BottomUpTransform.java:32)
	at org.apache.druid.sql.calcite.filtration.Filtration.transform(Filtration.java:86)
	at org.apache.druid.sql.calcite.filtration.Filtration.optimize(Filtration.java:113)
	at org.apache.druid.sql.calcite.rel.DruidQuery.toScanQuery(DruidQuery.java:995)
	at org.apache.druid.sql.calcite.rel.DruidQuery.computeQuery(DruidQuery.java:760)
	at org.apache.druid.sql.calcite.rel.DruidQuery.<init>(DruidQuery.java:167)
	at org.apache.druid.sql.calcite.rel.PartialDruidQuery.build(PartialDruidQuery.java:332)
	at org.apache.druid.sql.calcite.rel.DruidQueryRel.toDruidQuery(DruidQueryRel.java:97)
	at org.apache.druid.sql.calcite.rel.DruidQueryRel.toDruidQueryForExplaining(DruidQueryRel.java:109)
	at org.apache.druid.sql.calcite.rel.DruidRel.isValidDruidQuery(DruidRel.java:62)
	at org.apache.druid.sql.calcite.rule.DruidRules$DruidQueryRule.onMatch(DruidRules.java:137)
	at org.apache.calcite.plan.volcano.VolcanoRuleCall.onMatch(VolcanoRuleCall.java:212)
	... 44 more
```

I am working on a fix for it. It should return as if it was running on an empty dataset (since the filter matches nothing).

--- a/sql/src/main/java/org/apache/druid/sql/calcite/filtration/CombineAndSimplifyBounds.java
+++ b/sql/src/main/java/org/apache/druid/sql/calcite/filtration/CombineAndSimplifyBounds.java
@@ -118,48 +118,23 @@ public class CombineAndSimplifyBounds extends BottomUpTransform
    * Simplify BoundDimFilters that are children of an OR or an AND.
    *
    * @param children    the filters
-   * @param disjunction true for disjunction, false for conjunction
+   * @param disjunction true for OR, false for AND
    *
    * @return simplified filters
    */
   private static DimFilter doSimplify(final List<DimFilter> children, boolean disjunction)
   {
-    // Copy children list
+    // Copy the list of child filters. We'll modify the copy and eventually return it.
     final List<DimFilter> newChildren = Lists.newArrayList(children);
 
     // Group Bound filters by dimension, extractionFn, and comparator and compute a RangeSet for each one.
     final Map<BoundRefKey, List<BoundDimFilter>> bounds = new HashMap<>();
 
-    final Iterator<DimFilter> iterator = newChildren.iterator();
-    while (iterator.hasNext()) {
-      final DimFilter child = iterator.next();
-
-      if (child.equals(Filtration.matchNothing())) {
-        // Child matches nothing, equivalent to FALSE
-        // OR with FALSE => ignore
-        // AND with FALSE => always false, short circuit
-        if (disjunction) {
-          iterator.remove();
-        } else {
-          return Filtration.matchNothing();
-        }
-      } else if (child.equals(Filtration.matchEverything())) {
-        // Child matches everything, equivalent to TRUE
-        // OR with TRUE => always true, short circuit
-        // AND with TRUE => ignore
-        if (disjunction) {
-          return Filtration.matchEverything();
-        } else {
-          iterator.remove();
-        }
-      } else if (child instanceof BoundDimFilter) {
+    for (final DimFilter child : newChildren) {
+      if (child instanceof BoundDimFilter) {
         final BoundDimFilter bound = (BoundDimFilter) child;
         final BoundRefKey boundRefKey = BoundRefKey.from(bound);
-        List<BoundDimFilter> filterList = bounds.get(boundRefKey);
-        if (filterList == null) {
-          filterList = new ArrayList<>();
-          bounds.put(boundRefKey, filterList);
-        }
+        final List<BoundDimFilter> filterList = bounds.computeIfAbsent(boundRefKey, k -> new ArrayList<>());
         filterList.add(bound);
       }
     }
@@ -184,25 +159,13 @@ public class CombineAndSimplifyBounds extends BottomUpTransform
 
         if (rangeSet.asRanges().isEmpty()) {
           // range set matches nothing, equivalent to FALSE
-          // OR with FALSE => ignore
-          // AND with FALSE => always false, short circuit
-          if (disjunction) {
-            newChildren.add(Filtration.matchNothing());
-          } else {
-            return Filtration.matchNothing();
-          }
+          newChildren.add(Filtration.matchNothing());
         }
 
         for (final Range<BoundValue> range : rangeSet.asRanges()) {
           if (!range.hasLowerBound() && !range.hasUpperBound()) {
             // range matches all, equivalent to TRUE
-            // AND with TRUE => ignore
-            // OR with TRUE => always true; short circuit
-            if (disjunction) {
-              return Filtration.matchEverything();
-            } else {
-              newChildren.add(Filtration.matchEverything());
-            }
+            newChildren.add(Filtration.matchEverything());
           } else {
             newChildren.add(Bounds.toFilter(boundRefKey, range));
           }
@@ -210,8 +173,44 @@ public class CombineAndSimplifyBounds extends BottomUpTransform
       }
     }
 
+    // Finally: Go through newChildren, removing or potentially exiting early based on TRUE / FALSE marker filters.
     Preconditions.checkState(newChildren.size() > 0, "newChildren.size > 0");
-    if (newChildren.size() == 1) {
+
+    final Iterator<DimFilter> iterator = newChildren.iterator();
+    while (iterator.hasNext()) {
+      final DimFilter newChild = iterator.next();
+
+      if (Filtration.matchNothing().equals(newChild)) {
+        // Child matches nothing, equivalent to FALSE
+        // OR with FALSE => ignore
+        // AND with FALSE => always false, short circuit
+        if (disjunction) {
+          iterator.remove();
+        } else {
+          return Filtration.matchNothing();
+        }
+      } else if (Filtration.matchEverything().equals(newChild)) {
+        // Child matches everything, equivalent to TRUE
+        // OR with TRUE => always true, short circuit
+        // AND with TRUE => ignore
+        if (disjunction) {
+          return Filtration.matchEverything();
+        } else {
+          iterator.remove();
+        }
+      }
+    }
+
+    if (newChildren.isEmpty()) {
+      // If "newChildren" is empty at this point, it must have consisted entirely of TRUE / FALSE marker filters.
+      if (disjunction) {
+        // Must have been all FALSE filters (the only kind we would have removed above).
+        return Filtration.matchNothing();
+      } else {
+        // Must have been all TRUE filters (the only kind we would have removed above).
+        return Filtration.matchEverything();
+      }
+    } else if (newChildren.size() == 1) {
       return newChildren.get(0);
     } else {
       return disjunction ? new OrDimFilter(newChildren) : new AndDimFilter(newChildren);

--- a/sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java
+++ b/sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java
@@ -1892,6 +1892,28 @@ public class CalciteQueryTest extends BaseCalciteQueryTest
   }
 
   @Test
+  public void testGroupByNothingWithImpossibleTimeFilter() throws Exception
+  {
+    // Regression test for https://github.com/apache/incubator-druid/issues/7671
+
+    testQuery(
+        "SELECT COUNT(*) FROM druid.foo\n"
+        + "WHERE FLOOR(__time TO DAY) = TIMESTAMP '2000-01-02 01:00:00'\n"
+        + "OR FLOOR(__time TO DAY) = TIMESTAMP '2000-01-02 02:00:00'",
+        ImmutableList.of(
+            Druids.newTimeseriesQueryBuilder()
+                  .dataSource(CalciteTests.DATASOURCE1)
+                  .intervals(querySegmentSpec())
+                  .granularity(Granularities.ALL)
+                  .aggregators(aggregators(new CountAggregatorFactory("a0")))
+                  .context(TIMESERIES_CONTEXT_DEFAULT)
+                  .build()
+        ),
+        ImmutableList.of()
+    );
+  }
+
+  @Test
   public void testGroupByOneColumnWithLiterallyFalseFilter() throws Exception
   {
     testQuery(

